Timeline.cs

```diff
@@ -0,0 +1,507 @@
+/// </summary>
+/// <typeparam name="TValue">Value associated with a <see cref="System.DateTime"/>.</typeparam>
+internal class Timeline<TValue>:
+    ICollection<(DateTime Time, TValue Value)>,
+    IEquatable<Timeline<TValue>>,
+    ICloneable
+{
+    private List<(DateTime Time, TValue Value)> timeline;
+    bool ICollection<(DateTime Time, TValue Value)>.IsReadOnly { get => false; }
+
+    /// <summary>
+    /// The count of items.
+    /// </summary>
+    public int Count { get => timeline.Count; }
+
+    /// <summary>
+    /// Get all values associated with <paramref name="time"/>.
+    /// </summary>
+    /// <returns>Values associated with <paramref name="time"/>.</returns>
+    public TValue[] this[DateTime time]
+    {
+        get => GetValuesByTime(time);
+        set
+        {
+            var c = -1;
+            int i;
+            for (i = 0; i < Count; i++)
+            {
+                if (timeline[i].Time == time)
+                {
+                    if (c == value.Length)
+                        timeline.RemoveAt(i);
+                    else
+                        timeline[i] = (timeline[i].Time, value[++c]);
+                }
+            }
+            while (i < c)
+                Add(time, value[c]);
+        }
+    }
+
+    /// <summary>
+    /// Initialize the instance without items.
+    /// </summary>
+    internal Timeline()
+    {
+        timeline = new List<(DateTime, TValue)>();
+    }
+
+    /// <summary>
+    /// Initialize the instance with a <paramref name="time"/> and a <paramref name="value"/>.
+    /// </summary>
+    internal Timeline(DateTime time, TValue value)
+    {
+        timeline = new List<(DateTime, TValue)> { (time, value) };
+    }
+
+    /// <summary>
+    /// Initialize the instance with a <paramref name="value"/> associated with <see cref="System.DateTime.Now"/>.
+    /// </summary>
+    internal Timeline(params TValue[] value)
+    {
+        timeline = new List<(DateTime, TValue)>();
+        var now = DateTime.Now;
+        foreach (var v in value)
+            timeline.Add((now, v));
+    }
+
+    /// <summary>
+    /// Initialize the instance with a set of time and value.
+    /// </summary>
+    /// <param name="timeline"></param>
+    internal Timeline(params (DateTime, TValue)[] timeline)
+    {
+        this.timeline = timeline.ToList();
+        this.timeline = this.timeline.OrderBy(pair => pair.Time).ToList();
+    }
+
+    /// <summary>
+    /// Returns true if <paramref name="other"/> is equal to this timeline.
+    /// </summary>
+    /// <param name="other">Another <see cref="Timeline{TValue}"/> object.</param>
+    public bool Equals(Timeline<TValue> other)
+    {
+        return this == other;
+    }
+
+    /// <summary>
+    /// Returns a <see cref="System.Object"/> clone of the current instance.
+    /// </summary>
+    public object Clone()
```

Comment:
Could you remove cloning? I believe it's not essential from educational standpoint and complicates the overall design.